Improve DisconnectTip performance

sdaftuar opened the pull request:
During a reorg, we currently try to add transactions from each disconnected block back to the mempool, even though we are likely to find many of those transactions in the blocks that get connected.

This PR stores those transactions in a separate "disconnect pool" for later processing. This has a few side effects:

 * There should be many fewer transactions to re-add to the mempool, assuming the reorg is benign and most transactions appear on both forks, which means we'll do less work.

 * When we do add things to the mempool, we'll do it in one attempt rather than separately for each disconnected block, so mempool chain limits will be calculated and applied differently for reorgs than before. Overall it should be much more efficient, because we will have fewer calls chasing long chains in the mempool to keep track of ancestor and descendant state (both because we will call UpdateTransactionsFromBlock on a smaller set of transactions overall, and because the descendant caching it does isn't reused across calls, so it's more efficient to call UpdateTransactionsFromBlock once for all of a reorg's transactions than multiple times).

 * Storing the transactions in this new disconnect pool means we use more memory during a reorg. I've added a bound on how much memory can be used, and if we would exceed it we just try to keep the oldest transactions that are disconnected (the idea being that it's more important to ensure that older transactions that are reorged out have an opportunity to be re-mined, compared with newer ones). This could probably be improved in the future, since in such a (rare) scenario, we would probably want to throw out our current mempool in favor of being able to hold more older transactions, or we might want to re-read those transactions off disk to try to pick up everything that is now unconfirmed. But for now I made the memory bound large enough that this shouldn't practically occur.

sdaftuar committed:
Store disconnected block transactions outside mempool during reorg

Rather than re-add disconnected block transactions back to the mempool
immediately, store them in a separate disconnectpool for later processing,
because we expect most such transactions to reappear in the chain that is
still to be connected (and thus we can avoid the work of reprocessing those
transactions through the mempool altogether).

sdaftuar committed:
[qa] Relax assumptions on mempool behavior during reorg

Policy limits (such as chain limits and mempool total size) could reasonably
be enforced more aggressively during a reorg, so use resendwallettransactions
to repopulate the mempool to avoid mined blocks being too small, and increase
the chain limits from the default for this test.

This is in preparation for a change in mempool behavior during a reorg.

ryanofsky committed:
Add RecursiveDynamicUsage overload for std::shared_ptr

This simplifies a few usage expressions.

sdaftuar commented:
I just realized there's a potential issue here with the callbacks for conflicted transactions; in the case where a disconnected block transaction ends up conflicting with some tx in a block on the reorg, the call to ATMP will fail (because some input is spent) but not be reported via SyncTransaction() as a conflicted transaction.

sdaftuar commented:
Rebased, and removed WIP tag as this is now ready for review.

dcousens commented:
concept ACK,  will review

gmaxwell commented:
Needs rebase. utACK.

sdaftuar commented:
Rebased

ryanofsky commented:
utACK 37a90f1ad0b57de2ab4294dfa4a97d2e666dd479

PR makes sense, and nothing has changed except a few comments since my last review (for 4d70f47a94cc739e92f7d3f8f4f7b11cf7457ca9).

One thing I don't understand is why the pruning test needs to change when it sounds like you are saying the disconnect pool memory bound is large enough that losing transactions "shouldn't practically occur." Are you just updating the pruning test as a precaution, or are changes there needed because the chain it creates is unusual?

Also this PR has merge conflicts, but they are very minor. The validation.cpp conflict can be fixed by updating AcceptToMemoryPool arguments which changed in edded808fc4eee94c178e1779b90d1c87a08c23a (#9499).

dcousens reviewed (approved):
utACK 37a90f1

sdaftuar commented:
Rebased.

@ryanofsky The changes to pruning.py are because:
 * Policy limits get applied differently than before; we used to ignore some of the chain limits when adding transactions back to the mempool during a reorg (because we wouldn't look for in-mempool descendants until the end).
 * The pruning test is trying to build really big blocks in a number of places, to make the chain big enough to be pruned, and in particular in a few places it relies on the mempool being populated with disconnected transactions after a reorg (so changes to policy that would prevent that are problematic for the test).

So I mitigated this in part by increasing the chain limits and in part by trying to keep the mempool somewhat populated with a node's wallet transactions at a couple relevant places in the test.

[I actually briefly thought that because we now save the mempool in between restarts, we may not need the `resendwallettransactions()` calls I added (which I use to populate the mempool somewhat in a couple places), but I tested locally that removing them results in test failure (plausibly because the mined blocks aren't big enough, which I think is due to the reorg behavior change).  I do think this test needs to be overhauled now that many of the assumptions around mempool behavior have changed, but I think we should do that in another PR.]

sdaftuar commented:
Rebased after #9602.

ryanofsky commented on src/validation.cpp:
```diff
@@ -570,6 +570,44 @@ static bool IsCurrentForFeeEstimation()
+        if ((*it)->IsCoinBase() || !AcceptToMemoryPool(mempool, stateDummy, *it, false, NULL, NULL, true)) {
+            // If the transaction doesn't make it in to the mempool, remove any
+            // transactions that depend on it (which would now be orphans).
+            mempool.removeRecursive(**it);
```
In commit "Store disconnected block transactions outside mempool during reorg":

It looks like the new `MemPoolRemovalReason::REORG` argument for removeRecursive got dropped during the rebase. (Shame that we don't a have test that would catch this, and that the argument is optional to begin with.)

sdaftuar replied:
Thanks for catching this!  Fixed.

sipa replied:
I don't see `MemPoolRemovalReason::REORG` anywhere?

ryanofsky reviewed (commented):
utACK c8d4c5fa8f8d2375ea781523ae7619f1cefea6f9 if the MemPoolRemovalReason::REORG argument is added back.

Confirmed no other code changes since the previous rebase.

sdaftuar commented:
Rebased again to pick up the `pruning.py` fix from #9972

sdaftuar commented:
@sipa Thanks for reviewing (yikes, I think I must have done my second rebase on Friday using an older branch).  All the comments should be addressed in the last two fixup commits; let me know if I should go ahead and squash.

morcos commented on src/validation.cpp:
```diff
@@ -2402,6 +2444,10 @@ static bool ActivateBestChainStep(CValidationState& state, const CChainParams& c
 
+    // If any blocks were disconnected, disconnectpool may be non empty.  Add
+    // any disconnected transactions back to the mempool.
+    ReacceptToMemoryPool(disconnectpool);
```
nit: Any reason to put `ReacceptToMemoryPool` here and not under `if (fBlocksDisconnected)`?
It might make the logic slightly clearer.

sdaftuar replied:
Fixed

morcos commented:
fast review utACK

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -3697,7 +3751,9 @@ bool RewindBlockIndex(const CChainParams& params)
         }
-        if (!DisconnectTip(state, params, true)) {
+        DisconnectedBlockTransactions dummypool;
+        // fBare=true means nothing will actually be added to dummypool
```
Hmm, thats a super confusing API now....maybe either pass in a pointer to a DisconnectedBlockTransactions as both fBare and the pool?

sdaftuar replied:
Fixed.

TheBlueMatt commented on src/validation.h:
```diff
@@ -69,6 +69,8 @@ static const unsigned int DEFAULT_DESCENDANT_LIMIT = 25;
 /** Default for -mempoolexpiry, expiration time for mempool transactions in hours */
 static const unsigned int DEFAULT_MEMPOOL_EXPIRY = 336;
+/** Maximum kilobytes for transactions to store for processing during reorg */
+static const unsigned int MAX_DISCONNECTED_TX_POOL_SIZE = 75000;
```
nit: Yeesh, isnt that a bit high? Maybe like 10MB - if we reorg something larger than 2/3 blocks throwing away tons of txn is fine by me.

sdaftuar replied:
Dropping it to 20 (segwit blocks can be bigger).

TheBlueMatt reviewed (commented):
Strong concept ACK

dcousens commented:
Needs rebase?

sdaftuar commented on src/txmempool.h:
```diff
@@ -719,4 +725,81 @@ struct TxCoinAgePriorityCompare
+        for (auto const &tx : vtx) {
+            auto it = queuedTx.find(tx->GetHash());
+            if (it != queuedTx.end()) {
+                cachedInnerUsage -= RecursiveDynamicUsage(*it);
```
I'm going to leave this, since `removeEntry` takes an iterator into a different multi_index, and I'd rather removeForBlock be as fast as possible.

sdaftuar commented on src/validation.cpp:
```diff
@@ -2351,9 +2389,13 @@ static bool ActivateBestChainStep(CValidationState& state, const CChainParams& c
     while (chainActive.Tip() && chainActive.Tip() != pindexFork) {
-        if (!DisconnectTip(state, chainparams))
+        if (!DisconnectTip(state, chainparams, disconnectpool)) {
+            // The mempool will be inconsistent, but we're about to shutdown
```
Fixed.  I do think that a failure in DisconnectTip() should always cause shutdown, as we can no longer be sure we're in consensus, but no need to worry about that here.

sdaftuar commented:
First, I rebased https://github.com/sdaftuar/bitcoin/commits/fd-rebased.2 -> https://github.com/sdaftuar/bitcoin/commits/fd-rebased.3, no actual conflicts.

Then I added 4 fixup commits to address the nits mentioned so far.  I also added an assertion that a DisconnectedBlockTransactions object can't be destroyed if it's non-empty...  Not sure how good an idea that is, but I thought it might catch code bugs in the future related to mempool consistency?

sdaftuar commented:
Squashed all those commits down (https://github.com/sdaftuar/bitcoin/commits/fd-rebased.4 -> https://github.com/bitcoin/bitcoin/pull/9208/commits/5685657564e7a19635529eabea6af829a3a4996c)

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2379,7 +2429,7 @@ static bool ActivateBestChainStep(CValidationState& state, const CChainParams& c
         // Connect new blocks.
         BOOST_REVERSE_FOREACH(CBlockIndex *pindexConnect, vpindexToConnect) {
-            if (!ConnectTip(state, chainparams, pindexConnect, pindexConnect == pindexMostWork ? pblock : std::shared_ptr<const CBlock>(), connectTrace)) {
+            if (!ConnectTip(state, chainparams, pindexConnect, pindexConnect == pindexMostWork ? pblock : std::shared_ptr<const CBlock>(), connectTrace, disconnectpool)) {
```
We'll now assert() on L2445 when we return false if ConnectTip fails, instead of gracefully AbortNode()ing.

sdaftuar replied:
I don't know what the best answer here is.  I could call ReacceptToMemoryPool(), but doing so after we've invoked AbortNode() seems either annoying (trying to lookup UTXOs from disk when we're potentially having disk trouble seems bad), or potentially problematic (is it okay to trigger SyncTransaction() callbacks when our on-disk state is potentially corrupt?). 

One idea: check to see if fShutdownRequested is set, and if so, just wipe the disconnectpool and the mempool so that any observers of the mempool won't see a potentially inconsistent state (and if fShutdownRequested is not set, call ReacceptToMemoryPool() -- this should never happen).  Then the mempool saving won't work in the event of an AbortNode(), but maybe that's okay, we're probably out of disk space anyway.

@sipa Any suggestions on how to handle this?

TheBlueMatt replied:
For posterity: meeting conclusion, broadly, appeared to be that we should move forward with AbortNode() eventually migrating towards being used only for non-critical errors that result in a "we should shutdown soonish", and critical errors such as on-disk corruption should be an assert(false), ie that we should move forward here ensuring that the mempool is consistent when we return.

sdaftuar replied:
@TheBlueMatt I pushed up a commit that addresses the edge-cases better, I think -- it now will just erase from the mempool to become consistent again.  Also patches up the missing `mempool.removeForReorg` calls in a couple places, and moves all the reorg consistency handling into one function to get rid of some code duplication.

TheBlueMatt commented:
utACK 65a4dcc46f1ef942d52d8445cdc4a2ef385c816b, though please squash it with e4a0440fab759a0a3c2d4d03f9bd7463d84dde9a

sdaftuar commented:
Squashed 65a4dcc -> 688417b

morcos commented on src/txmempool.h:
```diff
@@ -690,4 +696,91 @@ class CCoinsViewMemPool : public CCoinsViewBacked
+    // Remove entries based on txid_index, and update memory usage.
+    void removeForBlock(const std::vector<CTransactionRef>& vtx)
+    {
+        for (auto const &tx : vtx) {
```
probably should shortcut to skip this loop if disconnectpool is empty which is the common case in connecttip

sdaftuar replied:
Agreed, fixed in 15b46d94b53af6429f808bb47ebc2a40da9dc791

morcos commented:
utACK 688417b modulo performance nit

sdaftuar commented:
Squashed 15b46d94b53af6429f808bb47ebc2a40da9dc791 -> 75b072df512442829bf288f82ed8ce0830512763

morcos commented:
utACK 75b072d

theuni commented on src/core_memusage.h:
```diff
@@ -63,4 +63,9 @@ static inline size_t RecursiveDynamicUsage(const CBlockLocator& locator) {
 
+template<typename X>
+static inline size_t RecursiveDynamicUsage(const std::shared_ptr<X>& p) {
+    return RecursiveDynamicUsage(*p) + memusage::DynamicUsage(p);
```
Watch the null deref:
```c++
return p ? memusage::DynamicUsage(p) + RecursiveDynamicUsage(*p) : 0;
```

sdaftuar replied:
Thanks, fixed in ea1b122

sdaftuar commented:
Fixed @theuni's nit in ea1b122 (https://github.com/sdaftuar/bitcoin/commits/fd-rebased.7), and squashed  in 2cb9509

sdaftuar commented:
Rebased again due to merge conflict in pruning.py

sdaftuar commented:
Anything else needed here?

gmaxwell reviewed (approved):
tested ACK (I previously utACKed, I know longer remember the code but I've used it a bunch and beat on it a bit)

TheBlueMatt commented:
re-utACK c1235e3f2dd5b01b63b020d1b8f7283e8badaf09

morcos commented:
re-utACK c1235e3
